client/state/jetpack/modules/reducer.js

```diff
@@ -16,53 +15,53 @@ import {
 import { combineReducers } from 'calypso/state/utils';
 
 const createItemsReducer = ( active ) => {
-	return ( state, { siteId, moduleSlug } ) => {
-		return merge( {}, state, {
-			[ siteId ]: {
-				[ moduleSlug ]: {
-					active,
-				},
+	return ( state, { siteId, moduleSlug } ) => ( {
+		...state,
+		[ siteId ]: {
+			...state[ siteId ],
+			[ moduleSlug ]: {
+				...state[ siteId ]?.[ moduleSlug ],
+				active,
 			},
-		} );
-	};
+		},
+	} );
 };
 
 const createItemsListReducer = () => {
-	return ( state, { siteId, modules } ) => {
-		return merge( {}, state, {
-			[ siteId ]: modules,
-		} );
-	};
+	return ( state, { siteId, modules } ) => ( {
+		...state,
+		[ siteId ]: modules,
+	} );
 };
 
 const createRequestsReducer = ( data ) => {
-	return ( state, { siteId, moduleSlug } ) => {
-		return merge( {}, state, {
-			[ siteId ]: {
-				[ moduleSlug ]: data,
-			},
-		} );
-	};
+	return ( state, { siteId, moduleSlug } ) => ( {
+		...state,
+		[ siteId ]: {
+			...state[ siteId ],
+			[ moduleSlug ]: data,
+		},
+	} );
 };
 
 const createModuleListRequestReducer = ( fetchingModules ) => {
-	return ( state, { siteId } ) => {
-		return merge( {}, state, {
-			[ siteId ]: {
-				fetchingModules,
-			},
-		} );
-	};
+	return ( state, { siteId } ) => ( {
+		...state,
+		[ siteId ]: {
+			fetchingModules,
+		},
+	} );
 };
 
 const createSettingsItemsReducer = () => {
 	return ( state, { siteId, settings } ) => {
 		let updatedState = state;
-		const moduleActivationState = pickBy( settings, ( settingValue, settingName ) => {
-			return get( state, [ siteId, settingName ] ) !== undefined;
-		} );
 
-		forEach( moduleActivationState, ( active, moduleSlug ) => {
+		const moduleActivationState = Object.entries( settings ).filter(
+			( [ settingName ] ) => state[ siteId ]?.[ settingName ] !== undefined
+		);
+
+		moduleActivationState.forEach( ( [ moduleSlug, active ] ) => {
```

Comment:
This would be a good case for `reduce`, reducing the array:
```js
updatedState = moduleActivationState.reduce( reducer, state );
```

Also to avoid cloning the `state` object too many times, change only the part of the tree that really needs changing:
```js
updatedState = {
  ...state,
  [ siteId ]: moduleActivationState.reduce( reducer, state[ siteId ] ),
}
```
Note that we're updating only one specific `state[ siteId ]` item, never multiple sites at once.